linalg/sparsemat.cpp

```diff
@@ -3995,4 +4045,29 @@ void SparseMatrix::Swap(SparseMatrix &other)
    mfem::Swap(isSorted, other.isSorted);
 }
 
+SparseMatrix::~SparseMatrix()
+{
+   Destroy();
+#ifdef MFEM_USE_CUDA_OR_HIP
+   if (useGPUSparse)
+   {
+      if (SparseMatrixCount==1)
+      {
+         if (handle)
+         {
+            cu_or_hip(sparseDestroy)(handle);
+            handle = nullptr;
+         }
+         if (dBuffer)
+         {
+            Cu_or_Hip(MemFree)(dBuffer);
+            dBuffer = nullptr;
+            bufferSize = 0;
+         }
+      }
+      SparseMatrixCount--;
```

Comment:
I just noticed this: I think we need to tweak the logic here a bit. On construction, this counter is incremented and `useGPUSparse` set to `true` if a CUDA or a HIP backend is enabled. If `useGPUSparse` is not changed after that, the code here correctly decrements the counter. However, if `UseGPUSparse(false)` was called, setting `useGPUSparse` to `false`, then this counter will not be decremented which is wrong. Basically, instead of checking for `useGPUSparse == true`, we need to do the same check as in `InitGPUSparse()` to see if a CUDA or a HIP backend is enabled.

Reply:
that makes sense!

Reply:
Pushed

I'm wondering if it would actually be better to release this cu/hipSPARSE stuff in an `atexit` callback so that you wouldn't init->cleanup->init->cleanup if you had only one spmat that you created then destroyed it multiple times

Reply:
I'm not sure how expensive the handle creation/destruction is. Do you know?

Reply:
For sure the first device wide call is truly expensive. I don't think the next are.

Reply:
I don't know. I'm happy with what we have now, it was just a thought, and we can revisit in the future if the performance impact comes up?

Reply:
Yes, we can look into this later, if needed.

Reply:
I agree that TPLs init and finalization should definitely be revisited, through the `Device` maybe.
`CuRand` for example has also such behavior and it would be nice to centralize the way of handling them.